### PR TITLE
refactor: Uniform environment variables management

### PR DESCRIPTION
**Pull request by zharinov**
Here is some possibility of breaking change.
It explicitly sets variables for each `exec()`, depending on `envVars` option and `trustLevel` setting.
With the current implementation, missing `env` argument defaults to `process.env` which won't be the case after merge.

Ref #2863

### CODE REVIEW

File: lib/util/docker/index.ts [hunk 1]
```diff
@@ -7,37 +9,55 @@ export interface DockerOptions {
   envVars?: Opt<Opt<string>[]>;
   cwd?: Opt<string>;
   tag?: Opt<string>;
-  cmdWrap?: Opt<string>;
+  preCommands?: Opt<string[]>;
```
Comment (rarkins) [inline lib/util/docker/index.ts]: Isn't this a separate change? Can we make it into a different PR to keep each PR as minimal as possible?

File: test/manager/bundler/artifacts.spec.ts [hunk 2]
```diff
@@ -30,13 +30,7 @@ describe('bundler.updateArtifacts()', () => {
...
 
     processEnv = process.env;
-    process.env = {
-      HTTP_PROXY: 'http://example.com',
-      HTTPS_PROXY: 'https://example.com',
-      NO_PROXY: 'localhost',
-      HOME: '/home/user',
-      PATH: '/tmp/path',
-    };
+    process.env = mockedEnv;
```
Comment (ViceIce) [inline test/manager/bundler/artifacts.spec.ts]: shouldn't we simply mock `getChildProcessEnv` from `lib/util/env` ?

Comment (zharinov) [reply test/manager/bundler/artifacts.spec.ts]: Some code relies on `process.env` before doing `exec`

Comment (ViceIce) [reply test/manager/bundler/artifacts.spec.ts]: So we should refactor all code directly referencing `process.env` to a single util function (seperate pr). Then we can test more easily.

**Comment by ViceIce** [comment]
depends on #5103

**Comment by ViceIce** [comment]
@zharinov needs deconflicting

**Comment by zharinov** [comment]
Can be merged as-is, or after #5114

File: test/util/exec.spec.ts [hunk 3]
```diff
@@ -3,107 +3,214 @@ import {
...
 
-const cmd = 'echo hello';
+describe(`Child process execution wrapper`, () => {
+  let processEnvOrig;
+  let trustLevelOrig;
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+    jest.resetModules();
+    processEnvOrig = process.env;
```
Comment (ViceIce) [inline test/util/exec.spec.ts]: Can we use the env mock here too?

Comment (zharinov) [reply test/util/exec.spec.ts]: No, here we need to test the lower level

Comment (ViceIce) [reply test/util/exec.spec.ts]: ok

**Comment by ViceIce** [comment]
needs deconflicting

**Review by viceice — DISMISSED** [review]

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/exec.ts [hunk 4]
```diff
@@ -4,13 +4,16 @@ import {
   ExecOptions as ChildProcessExecOptions,
 } from 'child_process';
 import { dockerCmd, DockerOptions } from './docker';
+import { getChildProcessEnv } from './env';
+import { defaultEnvVars } from './envVars';
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/util/exec.ts]: I tried so, but given up  because otherwise I ended up writing ugly boilerplate across lots of tests: 
```js
jest.mock('env.js', () => ({
  getChildProcessEnv: jest.fn(),
  defaultEnvVars: requireActual('env.js'),
})
```
instead of just `const env = mocked(_env);`

File: lib/util/exec.ts [hunk 5]
```diff
@@ -19,14 +22,41 @@ export interface ExecResult {
...
-  const pExecOptions = { ...options, encoding: 'utf-8' };
+  const pExecOptions = {
+    encoding: 'utf-8',
+    ...opts,
+  };
+  delete pExecOptions.envVars;
+  delete pExecOptions.docker;
+
+  const processEnv = getChildProcessEnv(envVars);
+  const isCustomEnv = !!opts.env || typeof opts.envVars !== 'undefined';
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/util/exec.ts]: Different trythiness semantics made first versions of my code work smoothly with existing tests/mocks, no there is no need for this.

Custom env is constructed if we (1) pass `env` option explicitly or (2) we have some additional vars to inherit from current `process.env`. Otherwise, we're using invironment obtained via `getChildProcessEnv(...)` call.

[the remark this answers is not in the document]
Comment (rarkins) [reply lib/util/exec.ts]: In what cases do we require this custom env logic and which do we not? I'm still trying to understand it.

File: lib/util/exec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/util/exec.ts]: Can we clear it up once we have completed the refactoring? I hope that env logic can be a mostly internal part of `exec` and not requiring multiple references.

File: lib/util/exec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/util/exec.ts]: https://github.com/renovatebot/renovate/blob/master/lib/manager/gomod/artifacts.ts#L54

After refactor, I'll pass `{ ..., env: { GCO_ENABLED: '0'}, ... }` as an exec option

File: lib/util/exec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/util/exec.ts]: Ah, got it. Just realized now we have no one place where it's being called without subsequent `exec()`.

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/util/exec.ts]: So I think answer is yes, let's do it with the next PR maybe?

File: lib/util/exec.ts [hunk 6]
```diff
@@ -4,13 +4,16 @@ import {
...
+import { getChildProcessEnv } from './env';
+import { basicEnvVars } from './env/basicVars';
 
 const pExec: (
   cmd: string,
   opts: ChildProcessExecOptions & { encoding: string }
 ) => Promise<ExecResult> = promisify(cpExec);
 
 export interface ExecOptions extends ChildProcessExecOptions {
+  inheritEnvVars?: string[];
```
Comment (rarkins) [inline lib/util/exec.ts]: Can you document this so that it's use is intuitive from using the file?
